File: db4oj/core/src/com/db4o/internal/fileheader/FileHeaderVariablePart1.java
/* Copyright (C) 2006  Versant Inc.  http://www.db4o.com */

package com.db4o.internal.fileheader;

import com.db4o.*;
import com.db4o.ext.*;
import com.db4o.internal.*;
import com.db4o.internal.activation.*;
import com.db4o.internal.slots.*;


/**
 * @exclude
 */
public class FileHeaderVariablePart1 {
    
    // The variable part format is:

    // (int) converter version
    // (byte) freespace system used
    // (int)  freespace address
    // (int) identity ID
    // (long) versionGenerator
	// (int) uuid index ID
	
    
    private static final int LENGTH = 1 + (Const4.INT_LENGTH * 4) + Const4.LONG_LENGTH + Const4.ADDED_LENGTH;
    
    protected final LocalObjectContainer _container;
    
    protected final SystemData _systemData;
    
    private int _id;
    
    public FileHeaderVariablePart1(LocalObjectContainer container, int id, SystemData systemData) {
        _id = id;
        _container = container;
        _systemData = systemData;
    }
    
    public byte getIdentifier() {
        return Const4.HEADER;
    }

    public int ownLength() {
        return LENGTH;
    }

    public void readThis(ByteArrayBuffer buffer) {
		if (Deploy.debug) {
		    buffer.readBegin(getIdentifier());
		}
        _systemData.converterVersion(buffer.readInt());
        _systemData.freespaceSystem(buffer.readByte());
        _systemData.freespaceAddress(buffer.readInt());
        _systemData.identityId(buffer.readInt());
        _systemData.lastTimeStampID(buffer.readLong());
        _systemData.uuidIndexId(buffer.readInt());
    }

    public void writeThis(ByteArrayBuffer buffer) {
		if (Deploy.debug) {
		    buffer.writeBegin(getIdentifier());
		}
        buffer.writeInt(_systemData.converterVersion());
        buffer.writeByte(_systemData.freespaceSystem());
        buffer.writeInt(_systemData.freespaceAddress());
        Db4oDatabase identity = _systemData.identity();
        buffer.writeInt(identity == null ? 0 : identity.getID(systemTransaction()));
        buffer.writeLong(_systemData.lastTimeStampID());
        buffer.writeInt(_systemData.uuidIndexId());
    }

	private Transaction systemTransaction() {
		return _container.systemTransaction();
	}
    
    public void readIdentity(LocalTransaction trans) {
        LocalObjectContainer file = trans.localContainer();
        Db4oDatabase identity = Debug4.staticIdentity ? 
        		Db4oDatabase.STATIC_IDENTITY : 
        		(Db4oDatabase) file.getByID(trans, _systemData.identityId());
        if (null != identity) {
        	// TODO: what?
        	file.activate(trans, identity, new FixedActivationDepth(2));
        	_systemData.identity(identity);
        }
    }

	public Runnable commit() {
		int length = _container.blockConverter().blockAlignedBytes(ownLength());
		if(_id == 0){
			_id = _container.allocatePointerSlot();
		}
		final Slot committedSlot = _container.readPointerSlot(_id);
		final Slot newSlot = allocateSlot(length);
		ByteArrayBuffer buffer = new ByteArrayBuffer(length);
		writeThis(buffer);
		_container.writeEncrypt(buffer, newSlot.address(), 0);
		return new Runnable(){
			public void run() {
				
				// FIXME: This is not transactional !!!
				_container.writePointer(_id, newSlot);
				
				if(committedSlot == null || committedSlot.isNull()){
					return;
				}
				_container.freespaceManager().freeTransactionLogSlot(committedSlot);
			}
		};
	}

	private Slot allocateSlot(int length) {
		Slot reusedSlot = _container.freespaceManager().allocateTransactionLogSlot(length);
		if(reusedSlot != null){
			return reusedSlot;
		}
		return _container.appendBytes(length);
	}

	public int id() {
		return _id;
	}

	public void read() {
    	Slot slot = _container.readPointerSlot(_id);
    	ByteArrayBuffer buffer = _container.readBufferBySlot(slot);
    	readThis(buffer);
	}

}

File: db4on/core/src/core/com/db4o/internal/fileheader/FileHeaderVariablePart1.cs
namespace com.db4o.@internal.fileheader
{
	/// <exclude></exclude>
	public class FileHeaderVariablePart1 : com.db4o.@internal.PersistentBase
	{
		private const int LENGTH = 1 + (com.db4o.@internal.Const4.INT_LENGTH * 4) + com.db4o.@internal.Const4
			.LONG_LENGTH + com.db4o.@internal.Const4.ADDED_LENGTH;

		private readonly com.db4o.@internal.SystemData _systemData;

		public FileHeaderVariablePart1(int id, com.db4o.@internal.SystemData systemData)
		{
			SetID(id);
			_systemData = systemData;
		}

		internal virtual com.db4o.@internal.SystemData SystemData()
		{
			return _systemData;
		}

		public override byte GetIdentifier()
		{
			return com.db4o.@internal.Const4.HEADER;
		}

		public override int OwnLength()
		{
			return LENGTH;
		}

		public override void ReadThis(com.db4o.@internal.Transaction trans, com.db4o.@internal.Buffer
			 reader)
		{
			_systemData.ConverterVersion(reader.ReadInt());
			_systemData.FreespaceSystem(reader.ReadByte());
			_systemData.FreespaceAddress(reader.ReadInt());
			ReadIdentity(trans, reader.ReadInt());
			_systemData.LastTimeStampID(reader.ReadLong());
			_systemData.UuidIndexId(reader.ReadInt());
		}

		public override void WriteThis(com.db4o.@internal.Transaction trans, com.db4o.@internal.Buffer
			 writer)
		{
			writer.WriteInt(_systemData.ConverterVersion());
			writer.Append(_systemData.FreespaceSystem());
			writer.WriteInt(_systemData.FreespaceAddress());
			writer.WriteInt(_systemData.Identity().GetID(trans));
			writer.WriteLong(_systemData.LastTimeStampID());
			writer.WriteInt(_systemData.UuidIndexId());
		}

		private void ReadIdentity(com.db4o.@internal.Transaction trans, int identityID)
		{
			com.db4o.@internal.LocalObjectContainer file = trans.i_file;
			com.db4o.ext.Db4oDatabase identity = (com.db4o.ext.Db4oDatabase)file.GetByID1(trans
				, identityID);
			file.Activate1(trans, identity, 2);
			_systemData.Identity(identity);
		}
	}
}
